diesel/src/sqlite/expression/functions.rs

```diff
@@ -422,3 +422,51 @@ define_sql_function! {
     #[sql_name = "json_pretty"]
     fn json_pretty_with_indentation<J: JsonOrNullableJsonOrJsonbOrNullableJsonb + MaybeNullableValue<Text>>(j: J, indentation: Nullable<Text>) -> J::Out;
 }
+
+#[cfg(feature = "sqlite")]
+define_sql_function! {
+    /// Returns 1 if the argument is well-formed JSON, or returns 0 if is not well-formed.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # include!("../../doctest_setup.rs");
+    /// #
+    /// # fn main() {
+    /// #     #[cfg(feature = "serde_json")]
+    /// #     run_test().unwrap();
+    /// # }
+    /// #
+    /// # #[cfg(feature = "serde_json")]
+    /// # fn run_test() -> QueryResult<()> {
+    /// #     use diesel::dsl::{sql, json_pretty_with_indentation};
+    /// #     use serde_json::{json, Value};
+    /// #     use diesel::sql_types::{Text, Json, Jsonb, Nullable};
+    /// #     let connection = &mut establish_connection();
+    ///
+    /// let version = diesel::select(sql::<Text>("sqlite_version();"))
+    ///         .get_result::<String>(connection)?;
+    ///
+    /// // Querying SQLite version should not fail.
+    /// let version_components: Vec<&str> = version.split('.').collect();
+    /// let major: u32 = version_components[0].parse().unwrap();
+    /// let minor: u32 = version_components[1].parse().unwrap();
+    /// let patch: u32 = version_components[2].parse().unwrap();
+    ///
+    /// if major > 3 || (major == 3 && minor >= 46) {
+    ///     /* Valid sqlite version, do nothing */
+    /// } else {
+    ///     println!("SQLite version is too old, skipping the test.");
+    ///     return Ok(());
+    /// }
+    ///
+    /// let result = diesel::select(json_valid::<Json>(json!({"x":35})))
+    ///     .get_result::<Integer>(connection)?;
+    ///
+    /// assert_eq!(1, result);
+    ///
+    /// # }
+    /// ```
+    #[sql_name = "json_valid"]
+    fn json_valid<J: JsonOrNullableJsonOrJsonbOrNullableJsonb + MaybeNullableValue<Text>>(j: J) -> Integer;
```

Comment:
```suggestion
    fn json_valid<J: JsonOrNullableJsonOrJsonbOrNullableJsonb + MaybeNullableValue<Bool>>(j: J) -> J::Out;
```

Diesel maps such flag integers to a boolean for sqlite so we use `Bool` as return type. That would usually be `-> Bool`. For this particular function sqlite returns `NULL` if a null value is passed in, which means we want to return `Nullable<Bool>` if the input value is nullable and `Bool` otherwise. The `MaybeNullableValue<Bool>` part takes care of that by transforming the generic parameter into the right variant. That variant then can be accessed via the `Out` associated type.